pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/HashUtils.java

```diff
@@ -44,8 +73,31 @@ public static Object hashPrimaryKey(PrimaryKey primaryKey, HashFunction hashFunc
         return new ByteArray(HashUtils.hashMD5(primaryKey.asBytes()));
       case MURMUR3:
         return new ByteArray(HashUtils.hashMurmur3(primaryKey.asBytes()));
+      case UUID:
+        return new ByteArray(HashUtils.hashUUID(primaryKey.getValues()));
       default:
         throw new IllegalArgumentException(String.format("Unrecognized hash function %s", hashFunction));
     }
   }
+
+  /**
+   * Concatenates the string representation of all values into a single byte array. Each element is prepended with its
+   * 4-byte length to ensure no collisions can happen. (sacrifice space efficiency for correctness)
+   */
+  private static byte[] concatenate(Object[] values) {
+    byte[][] allValueBytes = new byte[values.length][];
+    int totalLen = 0;
+    for (int j = 0; j < allValueBytes.length; j++) {
+      allValueBytes[j] = values[j] == null ? "null".getBytes(StandardCharsets.UTF_8)
```

Comment:
This is dangerous. This means we treat string "null" and real null the same

Reply:
Why is this dangerous?

Reply:
Say one record has string `"null"` and one record has real null, they should be identified as different primary key

Reply:
Is the default behavior to throw when there's a real null? I don't think we handle null in `PrimaryKey`: https://github.com/apache/pinot/blob/master/pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/PrimaryKey.java#L43

We can also throw here to be consistent with other functions.

Reply:
@Jackie-Jiang : any suggestions on the path forward here? I am open to any of the following options:

* Disambiguate between "null" and null. I can denote a null using a single 0 byte.
* Throw an error if there's a null in keeping with `PrimaryKey`.

Reply:
I checked the code, and `PrimaryKey` should never contain `null` (it will contain default null value instead). We can just let it throw when encountering `null`.